File: advent_php/advent_three.php
<?php

function solve_puzzle($memory) {
    if (is_array($memory)) {
        $memory = implode('', $memory);
    }
    
    
    $mul_pattern = 'mul\((\d{1,3}),(\d{1,3})\)';
    $do_pattern = 'do\(\)';
    $dont_pattern = "don't\(\)";
    
    $pattern = "/$mul_pattern|$do_pattern|$dont_pattern/";
    
    preg_match_all($pattern, $memory, $matches, PREG_SET_ORDER);
    
    $total = 0;
    $enabled = true;
    
    foreach ($matches as $match) {
        if ($match[0] == 'do()') {
            $enabled = true;
        } elseif ($match[0] == "don't()") {
            $enabled = false;
        } elseif ($enabled) {
            $x = intval($match[1]);
            $y = intval($match[2]);
            $total += $x * $y;
        }
    }
    
    return $total;
}


$memory = file('input.txt', FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

$result = solve_puzzle($memory);
echo "The sum of all enabled multiplication results is: $result\n";

?>

File: advent_py/advent_three.py
import re

def solve_puzzle(memory):
    if isinstance(memory, list):
        memory = ''.join(memory)
    
    # Patterns for instructions
    mul_pattern = r'mul\((\d{1,3}),(\d{1,3})\)'
    do_pattern = r'do\(\)'
    dont_pattern = r"don't\(\)"
    
    instructions = re.finditer(f'{mul_pattern}|{do_pattern}|{dont_pattern}', memory)
    
    total = 0
    enabled = True
    
    for instruction in instructions:
        if instruction.group() == 'do()':
            enabled = True
        elif instruction.group() == "don't()":
            enabled = False
        elif enabled:
            x, y = map(int, instruction.groups())
            total += x * y
    
    return total

# Example usage
with open('memory.txt', 'r') as file:
    corrupted_memory = file.readlines()

result = solve_puzzle(corrupted_memory)
print(f"The sum of all enabled multiplication results is: {result}")